File: backend/src/test/java/org/globe42/web/users/UserControllerMvcTest.java
package org.globe42.web.users;

import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.util.Collections;
import java.util.Optional;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.globe42.dao.NoteDao;
import org.globe42.dao.SpentTimeDao;
import org.globe42.dao.TaskDao;
import org.globe42.dao.UserDao;
import org.globe42.domain.User;
import org.globe42.test.GlobeMvcTest;
import org.globe42.web.security.CurrentUser;
import org.globe42.web.security.PasswordDigester;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.web.servlet.MockMvc;

/**
 * MVC test for {@link UserController}
 * @author JB Nizet
 */
@GlobeMvcTest(UserController.class)
public class UserControllerMvcTest {
    @MockBean
    private CurrentUser mockCurrentUser;

    @MockBean
    private UserDao mockUserDao;

    @MockBean
    private TaskDao mockTaskDao;

    @MockBean
    private NoteDao mockNoteDao;

    @MockBean
    private SpentTimeDao mockSpentTimeDao;

    @MockBean
    private PasswordGenerator mockPasswordGenerator;

    @MockBean
    private PasswordDigester mockPasswordDigester;

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    private MockMvc mvc;

    @Test
    public void shouldGetCurrentUser() throws Exception {
        User user = UserControllerTest.createUser(42L);
        when(mockCurrentUser.getUserId()).thenReturn(user.getId());
        when(mockUserDao.findNotDeletedById(user.getId())).thenReturn(Optional.of(user));

        mvc.perform(get("/api/users/me"))
           .andExpect(status().isOk())
           .andExpect(jsonPath("$.id").value(user.getId()))
           .andExpect(jsonPath("$.login").value(user.getLogin()))
           .andExpect(jsonPath("$.password").doesNotExist());
    }

    @Test
    public void shouldChangePasswordOfCurrentUser() throws Exception {
        User user = UserControllerTest.createUser(42L);
        when(mockCurrentUser.getUserId()).thenReturn(user.getId());
        when(mockUserDao.findNotDeletedById(user.getId())).thenReturn(Optional.of(user));

        ChangePasswordCommandDTO command = new ChangePasswordCommandDTO("newPassword");

        mvc.perform(put("/api/users/me/passwords")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsBytes(command)))
           .andExpect(status().isNoContent());
    }

    @Test
    public void shouldList() throws Exception {
        User user = UserControllerTest.createUser(42L);
        when(mockUserDao.findNotDeleted()).thenReturn(Collections.singletonList(user));

        mvc.perform(get("/api/users"))
           .andExpect(status().isOk())
           .andExpect(jsonPath("$[0].id").value(user.getId()))
           .andExpect(jsonPath("$[0].login").value(user.getLogin()))
           .andExpect(jsonPath("$[0].password").doesNotExist());
    }

    @Test
    public void shouldCreate() throws Exception {
        UserCommandDTO command = new UserCommandDTO("test", true);

        when(mockPasswordGenerator.generatePassword()).thenReturn("password");
        when(mockPasswordDigester.hash("password")).thenReturn("hashed");

        mvc.perform(post("/api/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsBytes(command)))
           .andExpect(status().isCreated())
           .andExpect(jsonPath("$.login").value(command.getLogin()))
           .andExpect(jsonPath("$.generatedPassword").value("password"));
    }

    @Test
    public void shouldUpdate() throws Exception {
        User user = UserControllerTest.createUser(42L);
        when(mockUserDao.findNotDeletedById(user.getId())).thenReturn(Optional.of(user));

        UserCommandDTO command = new UserCommandDTO("test", true);

        mvc.perform(put("/api/users/{userId}", user.getId())
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsBytes(command)))
           .andExpect(status().isNoContent());
    }

    @Test
    public void shouldDelete() throws Exception {
        User user = UserControllerTest.createUser(42L);
        when(mockUserDao.findNotDeletedById(user.getId())).thenReturn(Optional.of(user));

        mvc.perform(delete("/api/users/{userId}", user.getId()))
           .andExpect(status().isNoContent());
    }

    @Test
    public void shouldResetPassword() throws Exception {
        User user = UserControllerTest.createUser(42L);
        when(mockUserDao.findNotDeletedById(user.getId())).thenReturn(Optional.of(user));

        when(mockPasswordGenerator.generatePassword()).thenReturn("password");
        when(mockPasswordDigester.hash("password")).thenReturn("hashed");

        mvc.perform(post("/api/users/{userId}/password-resets", user.getId()))
           .andExpect(status().isCreated())
           .andExpect(jsonPath("$.login").value(user.getLogin()))
           .andExpect(jsonPath("$.generatedPassword").value("password"));
    }
}

File: backend/src/test/kotlin/org/globe42/web/users/UserControllerMvcTest.kt
package org.globe42.web.users

import com.fasterxml.jackson.databind.ObjectMapper
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import org.assertj.core.api.Assertions.assertThat
import org.globe42.dao.UserDao
import org.globe42.domain.User
import org.globe42.test.GlobeMvcTest
import org.globe42.web.security.CurrentUser
import org.globe42.web.security.PasswordDigester
import org.globe42.web.test.jsonValue
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.*

/**
 * MVC test for [UserController]
 * @author JB Nizet
 */
@GlobeMvcTest(UserController::class)
class UserControllerMvcTest(
    @Autowired private val mvc: MockMvc,
    @Autowired private val objectMapper: ObjectMapper
) {
    @MockkBean
    private lateinit var mockCurrentUser: CurrentUser

    @MockkBean
    private lateinit var mockUserDao: UserDao

    @MockkBean
    private lateinit var mockPasswordGenerator: PasswordGenerator

    @MockkBean
    private lateinit var mockPasswordDigester: PasswordDigester

    @Test
    fun `should get current user`() {
        val userId = 42L
        val user = createUser(userId)
        every { mockCurrentUser.userId } returns userId
        every { mockUserDao.findNotDeletedById(userId) } returns user

        mvc.get("/api/users/me").andExpect {
            status { isOk }
            jsonValue("$.id", user.id!!)
            jsonValue("$.login", user.login)
            jsonPath("$.password") { doesNotExist() }
        }
    }

    @Test
    fun `should change password of current user`() {
        val userId = 42L
        val user = createUser(userId)
        every { mockCurrentUser.userId } returns userId
        every { mockUserDao.findNotDeletedById(userId) } returns user
        every { mockPasswordDigester.hash("newPassword") } returns "hashedNewPassword"

        val command = ChangePasswordCommandDTO("newPassword")

        mvc.put("/api/users/me/passwords") {
            contentType = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsBytes(command)
        }.andExpect {
            status { isNoContent }
        }
    }

    @Test
    fun `should list`() {
        val user = createUser(42L)
        every { mockUserDao.findNotDeleted() } returns listOf(user)

        mvc.get("/api/users").andExpect {
            status { isOk }
            jsonValue("$[0].id", user.id!!)
            jsonValue("$[0].login", user.login)
            jsonPath("$[0].password") { doesNotExist() }
        }
    }

    @Test
    fun `should create`() {
        val command = UserCommandDTO("test", true)

        every { mockUserDao.existsByLogin(command.login) } returns false
        every { mockPasswordGenerator.generatePassword() } returns "password"
        every { mockPasswordDigester.hash("password") } returns "hashed"
        every { mockUserDao.save(any<User>()) } answers { arg<User>(0).apply { id = 42L } }

        mvc.post("/api/users") {
            contentType = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsBytes(command)
        }.andExpect {
            status { isCreated }
            jsonValue("$.login", command.login)
            jsonValue("$.generatedPassword", "password")
        }
    }

    @Test
    fun `should update`() {
        val userId = 42L
        val user = createUser(userId)

        val command = UserCommandDTO("test", true)

        every { mockUserDao.findNotDeletedByLogin(command.login) } returns null
        every { mockUserDao.findNotDeletedById(userId) } returns user

        mvc.put("/api/users/{userId}", user.id) {
            contentType = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsBytes(command)
        }.andExpect {
            status { isNoContent }
        }
    }

    @Test
    fun `should delete`() {
        val userId = 42L
        val user = createUser(userId)
        every { mockUserDao.findNotDeletedById(userId) } returns user

        mvc.delete("/api/users/{userId}", user.id).andExpect {
            status { isNoContent }
        }
    }

    @Test
    fun `should reset password`() {
        val userId = 42L
        val user = createUser(userId)
        every { mockUserDao.findNotDeletedById(userId) } returns user

        every { mockPasswordGenerator.generatePassword() } returns "password"
        every { mockPasswordDigester.hash("password") } returns "hashed"

        mvc.post("/api/users/{userId}/password-resets", user.id).andExpect {
            status { isCreated }
            jsonValue("$.login", user.login)
            jsonValue("$.generatedPassword", "password")
        }
    }

    @Test
    fun `should get profile`() {
        val userId = 42L
        val user = createUser(userId)
        every { mockCurrentUser.userId } returns userId
        every { mockUserDao.findNotDeletedById(userId) } returns user

        mvc.get("/api/users/me/profile").andExpect {
            status { isOk }
            jsonValue("$.login", user.login)
            jsonValue("$.admin", user.admin)
            jsonValue("$.email", user.email!!)
            jsonValue("$.taskAssignmentEmailNotificationEnabled", user.taskAssignmentEmailNotificationEnabled)
        }
    }

    @Test
    fun `should update profile`() {
        val userId = 42L
        val user = createUser(userId)
        every { mockCurrentUser.userId } returns userId
        every { mockUserDao.findNotDeletedById(userId) } returns user

        val command = ProfileCommandDTO(
            email = "jb@bar.com",
            taskAssignmentEmailNotificationEnabled = true
        )

        mvc.put("/api/users/me/profile") {
            contentType = MediaType.APPLICATION_JSON
            content = objectMapper.writeValueAsBytes(command)
        }.andExpect {
            status { isNoContent }
        }
        assertThat(user.email).isEqualTo(command.email)
        assertThat(user.taskAssignmentEmailNotificationEnabled).isEqualTo(command.taskAssignmentEmailNotificationEnabled)
    }
}
